File: quick-sort.py
import random


def quick_sort(arr, start=None, end=None):
    start = 0 if start is None else start
    end = len(arr) - 1 if end is None else end

    if start < end:
        pivot = partition(arr, start, end)
        quick_sort(arr, start, pivot-1)
        quick_sort(arr, pivot+1, end)

def partition(arr, start, end):
    pivot = arr[end]
    i = start - 1

    for j in range(start, end):
        if arr[j] < pivot:
            i += 1
            temp = arr[i]
            arr[i] = arr[j]
            arr[j] = temp

    temp = arr[i+1]
    arr[i+1] = arr[end]
    arr[end] = temp

    return i+1


arr = [i for i in range(100)]
random.shuffle(arr)
print(arr)
quick_sort(arr)
print(arr)

File: quicksort.js
function quicksort(array, p, r) {
    if (p < r) {
        var q = partition(array, p, r);

        quicksort(array, p, q-1);
        quicksort(array, q+1, r);
    }
}


function partition(array, p, r) {
    var x = array[r];
    var i = p-1;

    for (var j = p; j < r; ++j) {
        if (array[j] <= x) {
            i = i + 1;
            var tmp = array[i];
            array[i] = array[j];
            array[j] = tmp;
        }
    }

    var tmp = array[i+1];
    array[i+1] = array[r];
    array[r] = tmp;

    return i + 1;
}


var array =  [51, 41, 6, 63, 86, 96, 57, 78, 22, 19];

quicksort(array, 0, array.length - 1);

console.log(array);
